cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,277 @@
+sc[SelectorCache]
+pol[Network Policy]
+bpf[/BPF\]
+ipc[IPCache]
+nm[NameManager]
+dnsp{{DNS packet}}
+
+dnsp-- UpdateGeneratedDNS() -->nm
+sc-- IdentitySelectionUpdated() -->pol
+nm-- UpsertPrefix() -->ipc
+ipc-- UpdateIdentities() -->sc
+nm-- UpdateFQDNSelector() -->sc
+pol-- SyncPolicyMap() -->bpf
+```
+
+#### The SelectorCache
+
+The SelectorCache maintains a list of all "active" selectors for policies. If, for example, there is a network policy that grants access to label `team: database`, this selector will "live" in the SelectorCache. The policy will ask the SelectorCache for all identities that match this selector. Furthermore, the policy will "register" for incremental updates to this selector.
+
+#### NameManager
+
+The NameManager maintains two key lists:
+1. The mapping from DNS names to IPs
+2. The set of active FQDN selectors
+
+As DNS responses come in, the NameManager determines if the name is selected by any ToFQDN policies. If so, it inserts the IP in to the ipcache (if needed) and updates the relevant ToFQDN selectors.
+
+#### IPCache
+
+The IPCache stores the mapping from prefix to labels. It also allocates identities for labels, and manages the mapping from prefix -> numeric identity.
+
+### Incremental updates
+
+1. An identity is allocated for the IP:
+  a. `ipc.UpsertPrefixes()` is called, associating the complete set of CIDR labels with this IP address.
+  b. The IPCache determines the IP's full set of labels, allocates an identity, and informs the SelectorCache of this new IP
+2. The selector is updated:
+  a. The NameManager calls `sc.UpdateFQDNSelector()`, updating the set of selected IPs for this given FQDNSelector
+  b. The SelectorCache determines the complete set of identities selected
+  c. Any differences are pushed to network policies
+3. Endpoints are updated
+  Any pending incremental updates are batched up and applied to the endpoints' BPF PolicyMap. If needed, Envoy is also updated
+ 
+Only at this point can the DNS packet be returned to the waiting endpoint.
+
+## Performance bottlenecks
+
+### Many identities
+
+Some DNS names (e.g S3) will return many different IP addresses. S3 even has a TTL of 5 seconds, so the set of IPs will quickly be very broad. This causes a large number of identities to be allocated -- one per IP. These identities then need to plumbed through to all endpoints.
+
+### Label sizes
+
+CIDR labels, especially IPv6 labels, are large. For example, the full set of labels for the CIDR `2001:db8:cafe::cab:4:b0b:0/112` is 2KiB. Every extra-cluster identity always has the full set of CIDR identities calculated for it.
+
+----
+
+## Proposal
+
+This CFP proposes 3 key changes. Together, they preserve correctness while radically restructuring how prefixes are selected.
+
+1: Labels on prefixes propagate downwards. That is to say, if prefix `192.0.0.0/22` has label `foo:bar`, then prefix `192.0.2.0/24` also has label `foo:bar`, unless overridden. The IPCache aggregates labels accordingly.
```

Comment:
nit: GitHub doesn't recognize `1:` numbered lists, fix this.

Reply:
If I apply this more concretely to the current implementation, I think it translates to the following:

When the IPcache allocates a local Identity for an IP, it inspects all shorter prefixes in the IPCache and accumulates the labels in each shorter prefix until it gets the full set of CIDR (+ potentially other) labels. Then it allocates an Identity for that full label set.

Is that accurate? That's what I had in mind for the CIDRPolicy implementation originally.

Reply:
Yes, that's what I had in mind for the implementation as well. And, as you mention, there's nothing special about `cidr:` labels; we will someday have "arbitrary" labels on prefixes.

The interesting bit -- and not immediately relevant to this bullet point -- is that multiple CIDRs may share identities.